lib/stripe/credit_card_cloner.rb

```diff
@@ -1,38 +1,101 @@
+#   - a card (card_*) or payment_method (pm_*) stored (in a customer) in a platform account into
 #   - a payment method (pm_*) (in a new customer) in a connected account
 #
 # This is required when using the Stripe Payment Intents API:
 #   - the customer and payment methods are stored in the platform account
 #       so that they can be re-used across multiple sellers
-#   - when a card needs to be charged, we need to create it in the seller's stripe account
+#   - when a card needs to be charged, we need to clone (or find the clone)
+#       in the seller's stripe account
 #
-# We are doing this process every time the card is charged:
-#   - this means that, if the customer uses the same card on the same seller multiple times,
-#       the card will be created multiple times on the seller's account
-#   - to avoid this, we would have to store the IDs of every card on each seller's stripe account
-#       in our database (this way we only have to store the platform account ID)
+# To avoid creating a new clone of the card/customer each time the card is charged or
+# authorized (e.g. for SCA), we attach metadata { clone: true } to the card the first time we
+# clone it and look for a card with the same fingerprint (hash of the card number) and
+# that metadata key to avoid cloning it multiple times.
+
 module Stripe
   class CreditCardCloner
+    def find_or_clone(credit_card, connected_account_id)
+      if card = find_cloned_card(credit_card, connected_account_id)
+        card
+      else
+        clone(credit_card, connected_account_id)
+      end
+    end
+
+    private
+
     def clone(credit_card, connected_account_id)
       new_payment_method = clone_payment_method(credit_card, connected_account_id)
 
       # If no customer is given, it will clone the payment method only
-      return nil, new_payment_method.id if credit_card.gateway_customer_profile_id.blank?
+      return [nil, new_payment_method.id] if credit_card.gateway_customer_profile_id.blank?
 
       new_customer = Stripe::Customer.create({ email: credit_card.user.email },
                                              stripe_account: connected_account_id)
       attach_payment_method_to_customer(new_payment_method.id,
                                         new_customer.id,
                                         connected_account_id)
 
+      add_metadata_to_payment_method(new_payment_method.id, connected_account_id)
+
       [new_customer.id, new_payment_method.id]
     end
 
-    private
+    def find_cloned_card(card, connected_account_id)
+      matches = []
+      return matches unless fingerprint = fingerprint_for_card(card)
+
+      find_customers(card.user.email, connected_account_id).each do |customer|
+        find_payment_methods(customer.id, connected_account_id).each do |payment_method|
+          if payment_method_is_clone?(payment_method, fingerprint)
+            matches << [customer.id, payment_method.id]
+          end
+        end
+      end
```

Comment:
I would extract this into a method which returns the first element straight away. So you don't need the `matches` variable and you don't need to iterate further if you found a match:
```suggestion
      find_customers(card.user.email, connected_account_id).each do |customer|
        find_payment_methods(customer.id, connected_account_id).each do |payment_method|
          if payment_method_is_clone?(payment_method, fingerprint)
            return [customer.id, payment_method.id]
          end
        end
      end

      nil
```

Reply:
I think this is leftover from a previous (uncommitted) version of the method I'd written, where I was planning to send some sort of alert if more than one match was found; I subsequently changed things so that we should really only ever find one match.

Reply:
Addressed in 1cf7a56